crypto/crypto/src/multi_ed25519.rs

```diff
@@ -0,0 +1,547 @@
+        self.to_bytes()
+    }
+}
+
+impl Genesis for MultiEd25519PrivateKey {
+    fn genesis() -> Self {
+        let mut buf = [0u8; ED25519_PRIVATE_KEY_LENGTH];
+        buf[ED25519_PRIVATE_KEY_LENGTH - 1] = 1;
+        MultiEd25519PrivateKey(vec![Ed25519PrivateKey::try_from(buf.as_ref()).unwrap()], 1)
+    }
+}
+
+//////////////////////
+// PublicKey Traits //
+//////////////////////
+
+/// Convenient method to create a MultiEd25519PublicKey from a single Ed25519PublicKey.
+impl From<&Ed25519PublicKey> for MultiEd25519PublicKey {
+    fn from(ed_public_key: &Ed25519PublicKey) -> Self {
+        MultiEd25519PublicKey(vec![ed_public_key.clone()], 1u8)
+    }
+}
+
+/// Implementing From<&PrivateKey<...>> allows to derive a public key in a more elegant fashion.
+impl From<&MultiEd25519PrivateKey> for MultiEd25519PublicKey {
+    fn from(private_key: &MultiEd25519PrivateKey) -> Self {
+        let public_key = private_key
+            .0
+            .iter()
+            .map(|ed_private_key| ed_private_key.public_key())
+            .collect();
+        MultiEd25519PublicKey(public_key, private_key.1)
+    }
+}
+
+/// We deduce PublicKey from this.
+impl PublicKey for MultiEd25519PublicKey {
+    type PrivateKeyMaterial = MultiEd25519PrivateKey;
+}
+
+/// Get the hash output of the serialized MultiEd25519PublicKey.
+impl std::hash::Hash for MultiEd25519PublicKey {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        let encoded_pubkey = self.to_bytes();
+        state.write(&encoded_pubkey);
+    }
+}
+
+// Those are required by the implementation of hash above.
+impl PartialEq for MultiEd25519PublicKey {
+    fn eq(&self, other: &MultiEd25519PublicKey) -> bool {
+        self.to_bytes() == other.to_bytes()
+    }
+}
+
+impl Eq for MultiEd25519PublicKey {}
+
+impl TryFrom<&[u8]> for MultiEd25519PublicKey {
+    type Error = CryptoMaterialError;
+
+    /// Deserialize a MultiEd25519PublicKey. This method will also check for key and threshold
+    /// validity, and will only deserialize keys that are safe against small subgroup attacks.
+    fn try_from(bytes: &[u8]) -> std::result::Result<MultiEd25519PublicKey, CryptoMaterialError> {
+        let payload_length = bytes.len();
+        // Special case for single keys to speed up deserialization.
+        if payload_length == ED25519_PUBLIC_KEY_LENGTH {
+            return Ed25519PublicKey::try_from(bytes)
+                .map(|public_key| MultiEd25519PublicKey(vec![public_key], 1u8));
+        }
+        let threshold = get_threshold(
+            payload_length,
+            ED25519_PRIVATE_KEY_LENGTH,
+            bytes[payload_length - 1],
+        )?;
+
+        let mut public_keys: Vec<Ed25519PublicKey> = vec![];
+        let multi_key = bytes
+            .chunks(ED25519_PUBLIC_KEY_LENGTH)
+            .filter(|chunk| chunk.len() == ED25519_PUBLIC_KEY_LENGTH)
+            .try_for_each(|key_bytes| deserialize_and_push(key_bytes, &mut public_keys));
```

Comment:
See https://github.com/libra/libra/pull/2723/files#r385483089

Reply:
@huitseeker is the recommended solution fail-fast? I thought that's the purpose of `try_for_each`